### PR TITLE
SMS-435 Add sms_number and sms_consent to Subscribers API

### PR DESCRIPTION
**Pull request by jeremyhaberman**
Add two attributes to the public Subscriber model for the main /v2/:account_id/subscribers/* routes:
* `sms_number`
* `sms_consent`

We are intentionally leaving out the doc updates for the /campaigns/:campaign_id/subscribers routes.

### CODE REVIEW

**Comment by gabelimon** [comment]
This looks great. We should double-check this when the other code is finalized otherwise I'd approve

**Review by gabelimon — DISMISSED** [review]

File: source/includes/rest/_subscribers.md [hunk 1]
```diff
@@ -16,6 +16,8 @@
   "zip": "90210",
   "country": "US",
   "phone": "555-555-5555",
+  "sms_phone_number": "+16125551212",
```
Comment (gabelimon) [inline source/includes/rest/_subscribers.md:19]: @jacobpmeyer is the phone number exposed like in this payload??

**Comment by gabelimon** [comment]
Screenshots of the changes:
Screen Shot 2022-02-01 at 1 10 25 PM
Screen Shot 2022-02-01 at 1 08 55 PM
Screen Shot 2022-02-01 at 1 11 37 PM

File: source/includes/rest/_subscribers.md (hunk 1 above)
Comment (jacobpmeyer) [reply source/includes/rest/_subscribers.md:19]: I thought that was the case, but I don't think we ever discussed it.

Comment (jacobpmeyer) [inline source/includes/rest/_subscribers.md:19]: This will need to be `"sms_number"`

File: source/includes/rest/_subscribers.md [hunk 2]
```diff
@@ -117,6 +119,14 @@ On January 7, 2019 we renamed Subscribers to People to better represent who you
       <td><code>phone</code></td>
       <td>The subscriber's primary phone number.</td>
     </tr>
+    <tr>
+      <td><code>sms_phone_number</code></td>
```
Comment (jacobpmeyer) [inline source/includes/rest/_subscribers.md:123]: Here and line 355 -> sms_number

**Review by jacobpmeyer — DISMISSED** [review]
Looks good. Just need the couple of changes to `sms_number`

File: source/includes/rest/_workflows.md [hunk 3]
```diff
@@ -450,6 +450,14 @@ If the workflow is not active, the subscriber will not be added to the workflow.
       <td><code>phone</code></td>
       <td>Optional. The subscriber's primary phone number.</td>
     </tr>
+    <tr>
+      <td><code>sms_phone_number</code></td>
```
Comment (jeremyhaberman) [inline source/includes/rest/_workflows.md:454]: This instance of `sms_phone_number` should change to `sms_number`.

File: source/includes/rest/_subscribers.md [hunk 4]
```diff
@@ -341,6 +351,14 @@ If you need to create or update a collection of subscribers at once, use our [ba
       <td><code>phone</code></td>
       <td>Optional. The subscriber's primary phone number.</td>
     </tr>
+    <tr>
+      <td><code>sms_number</code></td>
+      <td>Optional. String. The subscriber's mobile phone number in E.164 formatting. E.g. <code>"+16125551212"</code>. Only US-based numbers are supported at this time.</td>
+    </tr>
+    <tr>
+      <td><code>sms_consent</code></td>
+      <td>Optional. Boolean. Required if <code>sms_number</code> is not <code>null</code>. <code>true</code> if the person has granted consent to receive marketing and other communication via SMS; <code>false</code> otherwise. If you’re unsure whether or not you have gained legal SMS consent, check out our <a href="https://my.drip.com/docs/manual/sms/compliance-manage-sms-compliance">TCPA requirements article</a>.</td>
```
Comment (jeremyhaberman) [inline source/includes/rest/_subscribers.md:360]: I think we should remove the note about it being required. It's not anymore. We'll still create the SMS person for the number, but they'll be opted out.

Also, if they don't include `sms_consent` or set its value to `null`:
* Internally, we'll consider it `nil` by default but in some ways treat it the same as `false` (we'll still create the SMS person)
* Externally, it'll look like it was set to false (`json.sms_consent subscriber.sms_opted_in?`)

This is a little weird, and I'm not sure how to best describe that behavior in the contract. Maybe we just say that the default is `false` and `null` will be treated the same as `false`?

File: source/includes/rest/_subscribers.md [hunk 5]
```diff
@@ -117,6 +119,14 @@ On January 7, 2019 we renamed Subscribers to People to better represent who you
...
       <td>The subscriber's primary phone number.</td>
     </tr>
+    <tr>
+      <td><code>sms_number</code></td>
+      <td>The subscriber's mobile phone number in E.164 formatting.</td>
+    </tr>
+    <tr>
+      <td><code>sms_consent</code></td>
+      <td>A boolean describing whether the person has granted consent to receive marketing and other communication via SMS.</td>
+    </tr>
```
Comment (jeremyhaberman) [inline source/includes/rest/_subscribers.md:129]: We should update these descriptions to match the others.

It's too bad we have to duplicate the descriptions in three spots. Wonder if there's a simple, low-effort way to resolve this.

**Comment by jeremyhaberman** [comment]
@gabelimon, it's a little strange that I'm technically the author on this (from Github's perspective), but I think you took over ownership of the ticket and the work. I'm happy to work on the updates I mentioned, but I didn't want to step on your toes.

Please let me know how you'd prefer to proceed.

**Comment by jeremyhaberman** [comment]
I'm assuming we want to get this deployed for Sleeknote. True story?

**Comment by jeremyhaberman** [comment]
I talked to Gabe. I'm going to work on the changes.

**Comment by jeremyhaberman** [comment]
@gabelimon, @jacobpmeyer, @morganbird, this is ready to be reviewed again.

**Review by gabelimon — DISMISSED** [review]

**Review by jacobpmeyer — DISMISSED** [review]

File: source/includes/rest/_subscribers.md [hunk 6]
```diff
@@ -117,6 +119,14 @@ On January 7, 2019 we renamed Subscribers to People to better represent who you
       <td><code>phone</code></td>
       <td>The subscriber's primary phone number.</td>
     </tr>
+    <tr>
+      <td><code>sms_number</code></td>
+      <td>Optional. String. The subscriber's mobile phone number in E.164 formatting. E.g. <code>"+16125551212"</code>. Only US-based numbers are supported at this time.</td>
```
Comment (morganbird) [inline source/includes/rest/_subscribers.md:124]: Do we want to say explicitly that this will return a validation error?

**Review by morganbird — DISMISSED** [review]

File: source/includes/rest/_subscribers.md (hunk 6 above)
Comment (jeremyhaberman) [reply source/includes/rest/_subscribers.md:124]: I'm torn. I like the idea. But it doesn't seem like we do that for the other attributes. We have this catch-all errors page:
https://developer.drip.com/#validation-errors

But I also just noticed something. We need to add `sms_error` to that list. Example response with a bad number:

```json
{
  "errors": [
    {
      "code": "sms_error",
      "attribute": "sms_number",
      "message": "SMS number SMS number must be from an SMS supported country"
    },
    {
      "code": "sms_error",
      "attribute": "sms_number",
      "message": "SMS number SMS number is not valid mobile number"
    }
  ]
}

And I see what we're duplicating "SMS number" in the message.

I'll take care of these two issues.

**Comment by jeremyhaberman** [comment]
@gabelimon, @jacobpmeyer, @morganbird: My new commit reset your approval.

I added `sms_error` to the Errors page, but I left the error details out of the attribute descriptions to be consistent with the other attributes.

File: source/includes/rest/_subscribers.md (hunk 6 above)
Comment (morganbird) [reply source/includes/rest/_subscribers.md:124]: Probably best to be consistent for now

**Review by morganbird — APPROVED** [review]

**Pull request merged**